packages/core/src/services/documents/runDocumentVersion.ts

```diff
@@ -1,32 +1,80 @@
-import { Chain, createChain } from '@latitude-data/compiler'
+import path from 'path'
+
+import {
+  Chain,
+  createChain,
+  readMetadata,
+  Document as RefDocument,
+} from '@latitude-data/compiler'
 import {
   ChainEvent,
   ChainEventTypes,
-  DocumentVersion,
+  Commit,
   LATITUDE_EVENT,
   PROVIDER_EVENT,
   ProviderApiKey,
 } from '$core/browser'
-import { findWorkspaceFromDocument } from '$core/data-access'
-import { Result } from '$core/lib'
+import { findWorkspaceFromCommit } from '$core/data-access'
+import { NotFoundError, Result } from '$core/lib'
 import { streamToGenerator } from '$core/lib/streamToGenerator'
-import { ProviderApiKeysRepository } from '$core/repositories'
+import {
+  DocumentVersionsRepository,
+  ProviderApiKeysRepository,
+} from '$core/repositories'
 import { z } from 'zod'
 
 import { ai } from '../ai'
 
 export async function runDocumentVersion({
-  document,
+  documentUuid,
+  commit,
   parameters,
 }: {
-  document: DocumentVersion
+  documentUuid: string
+  commit: Commit
   parameters: Record<string, unknown>
 }) {
-  const workspace = await findWorkspaceFromDocument(document)
+  const workspace = await findWorkspaceFromCommit(commit)
   if (!workspace) throw new Error('Workspace not found')
 
   const scope = new ProviderApiKeysRepository(workspace.id)
-  const chain = createChain({ prompt: document.content, parameters })
+  const documentScope = new DocumentVersionsRepository(workspace.id)
+  const docs = await documentScope.getDocumentsAtCommit(commit)
+  if (docs.error) return Result.error(docs.error)
+
+  const document = docs.value.find((d) => d.documentUuid === documentUuid)
+  if (!document) return Result.error(new NotFoundError('Document not found'))
+
+  let resolvedContent: string = document.resolvedContent!
+
+  if (document.resolvedContent === undefined || commit.mergedAt === null) {
+    const referenceFn = async (
+      refPath: string,
+      from?: string,
+    ): Promise<RefDocument | undefined> => {
+      const fullPath = path
+        .resolve(path.dirname(`/${from ?? ''}`), refPath)
+        .replace(/^\//, '')
+
+      const document = docs.value.find((d) => d.path === fullPath)
+      if (!document) return undefined
+
+      return {
+        path: fullPath,
+        content: document.content,
+      }
+    }
+
+    const metadata = await readMetadata({
+      prompt: document.content,
+      fullPath: document.path,
+      referenceFn,
+    })
+
+    resolvedContent = metadata.resolvedPrompt
+  }
+
+  const chain = createChain({ prompt: resolvedContent, parameters })
```

Comment:
So you have to pass resolved prompt to the chain? Seems unexpected. I'd expect the chain to receive the reference function and the prompt and the parameters and do all the shenanigans internally. 

(Totally forgot about references lol)

Reply:
Yeah, I get this is not the expected usage. But chains require having references resolved, and for that it needs a `referenceFn` to know how to resolve them. We could add it to the `createChain` function, if that makes it easier to understand.